crates/service-protocol/src/discovery.rs

```diff
@@ -360,6 +393,24 @@ impl ServiceDiscovery {
     }
 }
 
+fn is_possible_h1_only_error(err: &DiscoveryError) -> bool {
+    if let DiscoveryError::Client(ServiceClientError::Http(
+        restate_service_client::http::HttpError::Hyper(err),
+    )) = err
+    {
+        // there is a race condition in hyper; sometimes we see the proper h2 FRAME_SIZE_ERROR
+        // other times we just see a generic 'canceled' error
+        return err.is_canceled()
+            || err
+                .source()
+                .and_then(|err| err.downcast_ref::<h2::Error>())
+                .and_then(|err| err.reason())
+                == Some(h2::Reason::FRAME_SIZE_ERROR);
```

Comment:
Is it somewhere documented how an HTTP1.1 server will react when connecting via an HTTP2 client?

Reply:
In theory there is HTTP 505. The cloudflare worker dev server that motivates this PR returns a HTTP 501 not implemented - but this is just because it tries to read the first few bytes to determine the method, and finds a method it doesn't understand. In any case, an h2 client is not going to understand this response. We can however be fairly confident that http1.1 responses will trigger these frame size errors in the rust h2 library, I suppose? perhaps other types of http1.1 server will do different things eg just close the connection, and we probably wouldn't detect it

Reply:
I actually don't know how h2 will behave. Looking at the http/2 error codes, there is also `HTTP_1_1_REQUIRED` (https://github.com/hyperium/h2/blob/master/src/frame/reason.rs#L60) which seems somehow more fitting. But I am not sure whether an HTTP1.1 server will respond with this error code.

Reply:
this would require the server to be able to send an h2 frame, which is surprising!